nitorch/tools/affine_reg.py

```diff
@@ -0,0 +1,961 @@
+        grid = identity(dm,
+            dtype=torch.float32, device=opt['device'], jitter=jitter, step=samp)
+    else:
+        # Use mean-space, dat_samp is an image with zeros.
+        dm = dat  # tuple
+        grid = identity(dm,
+            dtype=torch.float32, device=opt['device'], jitter=jitter, step=samp)
+        dat = torch.zeros(dm, dtype=torch.float32, device=opt['device'])
+    # Sub-sampled fixed image
+    dat_samp = dat[::samp[0], ::samp[1], ::samp[2]]
+
+    return dat_samp, grid
+
+
+def _get_mean_space(mats, dims):
+    """Compute mean space from images' field of views.
+
+    Parameters
+    ----------
+    mats :  [N,] tensor_like
+        List of affine matrices.
+    dims :  [N,] tensor_like
+        List of image dimensions.
+
+    Returns
+    ----------
+    mats : (4, 4) tensor_like
+        Mean affine matrix.
+    dims : (3,) tuple
+        Mean image dimensions.
+
+    """
+    # Copy matrices and dimensions to torch tensors
+    dtype = mats[0].dtype
+    device = mats[0].device
+    N = len(mats)
+    Mat = torch.zeros((4, 4, N), dtype=dtype, device=device)
+    Dim = torch.zeros((3, N), dtype=dtype, device=device)
+    for i in range(N):
+        Mat[..., i] = mats[i].clone()
+        Dim[..., i] = dims[i].clone()
+    # Compute mean-space
+    dim, mat, _ = mean_space(Mat, Dim)
+    dim = tuple(dim.cpu().int().tolist())
+    mat = mat.type(dtype)
+
+    return mat, dim
+
+
+def _hist_2d(img0, img1, fwhm=7, mx_int=255):
+    """Make 2D histogram.
+
+    Pre-requisites:
+    * Images same size
+    * Images same min and max intensities
+
+    # Naive method for computing a 2D histogram
+    h = torch.zeros((mx_int + 1, mx_int + 1))
+    for n in range(num_vox):
+        h[img0[n], mg1[n]] += 1
+
+    Parameters
+    ----------
+    img0 : (X, Y, Z) tensor_like
+        First image volume.
+    img1 : (X, Y, Z) tensor_like
+        Second image volume.
+    fwhm : float, default=7
+        Full-width at half max of Gaussian kernel, for smoothing
+        histogram.
+    mx_int : float, default=255
+        Max intensity of input images.
+
+    Returns
+    ----------
+    H : (mx_int + 1, mx_int + 1) tensor_like
+        Joint intensity histogram.
+
+    """
+    fwhm = (fwhm,) * 2
+
+    # Convert each 'coordinate' of intensities to an index
+    # (replicates the sub2ind function of MATLAB)
+    img0 = img0.flatten().floor()
+    img1 = img1.flatten().floor()
+    sub = torch.stack((img0, img1),
+                      dim=1)  # (num_vox, 2)
+    to_ind = torch.tensor((1, mx_int + 1),
+                          dtype=sub.dtype, device=img0.device)[..., None]  # (2, 1)
+    ind = torch.tensordot(sub, to_ind, dims=([1], [0]))  # (nvox, 1)
+
+    # Build histogram H by adding up counts according to the indicies in ind
+    H = torch.zeros(mx_int + 1, mx_int + 1, device=img0.device, dtype=ind.dtype)
+    H.put_(ind.long(), torch.ones(1, device=img0.device, dtype=ind.dtype).expand_as(ind), accumulate=True)
```

Comment:
Note: I didn't know that put had an `accumulate` mode! We could have used that to implement push (in a slightly less efficient way).

Reply:
Okay, interesting. The annoying thing is that it requires long, and that conversion is a bit slow...

What do you think about the _hist_2d implementation otherwise? Don't you have some MI implementation that we could potentially use, as well?

Reply:
I think I have commented elsewhere: your implementation is probably more efficient (in time and memory) than mine. I don't think it's an issue to have to cast the index tensor. It would be nice to test if we can backpropagate through your implementation.